module/zfs/zfs_vnops.c

```diff
@@ -3657,6 +3792,9 @@ zfs_rename(struct inode *sdip, char *snm, struct inode *tdip, char *tnm,
 		error = zfs_link_create(tdl, szp, tx, ZRENAMING);
 		if (error == 0) {
 			szp->z_pflags |= ZFS_AV_MODIFIED;
+			if (tdzp->z_pflags & ZFS_PROJINHERIT &&
+			    S_ISDIR(ZTOI(szp)->i_mode))
+				szp->z_pflags |= ZFS_PROJINHERIT;
```

Comment:
In fact, it is difficult to guarantee that. For example, dir_a/dir_b/dir_c/file_d, firstly we set file_d's project ID as 100, then set dir_c's project ID as 200, and set ZFS_PROJINHERIT on dir_c. But the file_d has different project ID as dir_c. In theory, we can scan the whole dir_c when set ZFS_PROJINHERIT flag on it, but from the implementation view, its overhead may be too high to be accepted. So currently, we only guarantee that all the new created/renamed/linked file will inherit the parent's project ID, but not control the existing ones. I have to say that the "tree quota" still need more work.